### ISSUE TITLE
Deduce member/method types for a class that explicitly implements an interface.

### ISSUE THREAD

**Issue by Veetaha** [Microsoft/TypeScript#30200]
## Search Terms

"typescript interface implements deduction", "deduction implements"

## Suggestion
I suggest leveraging type deduction based on the declared `interface` when using the `implements` construct.
```ts
interface Foo { /* typed members/methods */ }
class Bar implements Foo { /* default types for members/methods are deduced */ }
```
I am surprised to have to retype all the members/methods types for a class even, though, it explicitly `implements` an interface with the necessary typing information.

I suggest that if you don't specify the types for method arguments and properties, they should be inferred from the interfaces, this class `implements`. If the type is not specified and there exists no interface with the proper declaration, then a compile-time error should be raised or given `any` type (depends on the `noImplicitAny` configuration flag).

My proposal is very straightforward, reduces code bloat and subdues to the [DRY concept](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

### Deducing method arguments and return types.

This is pretty simple.

Say we have the following code:
```ts
interface Interface {
    method(num: number, str?: string, ...someArgsIDontUse: boolean[]): boolean;
}

class Class implements Interface /*, Interface2, ..., InterfaceN */ {
    method(num, str) {
        return Boolean(num || str);
    }
}
```
In this case `method()` arguments `num` and `str` should have `number` and `string | undefined` types by default.

There already exists a mechanism to do arguments and return type deduction for object literals:
```ts
const obj: Interface /* & Interface2 & ... & InterfaceN */ = {
    method(num, str) {
        return Boolean(num || str);
    }
}
```
So why not integrating this functionality into class definitions?
### Deducing properties types.
Let a class have implicit property declaration for those members, that are declared in `Interface`, if they are not declared in the class definition at all.
```ts
interface Interface {
    property: string;
}

class Class implements Interface  /*, Interface2, ..., InterfaceN */  {
    constructor(arg: string) {
        this.property = arg;
    }
}
```
Here, when you declare no properties on the class, there exists an implicit declaration according to the implemented interfaces:
```ts
class Class implements Inteface /*, Interface2, ..., InterfaceN */  {
    property: string; /* implicit */
    /* other properties declared on Interface2, ..., InterfaceN */
}
```
Of course nothing must stop you from manually specifying the property type or using getters/setters for it. 

Shorthand class property initialization should also be typed by default:
```ts
class Class implements Interface {
    constructor(public property) {} /* property type string is deduced */
}
```

Checks for `strictPropertyInitialization` should give you an error message if you don't set the value for the properties deduced from `Interface`.
If you don't initialize the property, you should put an assertion the following way:
```ts
class Class implements Interface {
    property!; /* repeating string type is unnecessary */
    constructor() {}
}
```

## Use Cases

Besides making your code more succinct without losing type safety, this should solve such issues as #14790.

So when you declare a property on an interface to have some literal type, you won't have to repeat that literal two to three times.
Before:
```ts
interface Interface {
    property: "literal"
}
class Class implements Interface {
    property: "literal" = "literal"; 
    /* property  = "literal" as "literal"; */
    /* property!: "literal"; */
}
```
After:
```ts
class Class implements Interface {
    property = "literal"; /* property type is deduced to be literal from Interface */
    /* property!; */
    /* All declarations from "Before" are possible, but obviously redundant */
}
```
Of course you may prepend `public` keyword for declared properties if you want to be that explicit.

## Examples
Before:
```ts
interface Identifiable {
    id: string;
}
interface Named {
    name: string;
}
interface Foreachable<T> {
   foreEach(...args: Parameters<T[]['forEach']>): void;
}

class User implements Identifiable, Named {
    id: string;
    constructor(public name: string) {
        this.id = Utils.generateUUID();
    }
}
class Schedule implements Identifiable, Foreachable<string> {
    id!: string;
    private events = new Array<string>();

    forEach(...args: Parameters<string[]['forEach']>) {
        return this.events.forEach(...args);
    }

}

```
After:
```ts
interface Identifiable {
    id: string;
}
interface Named {
    name: string;
}
interface Foreachable<T> {
   foreEach(...args: Parameters<T[]['forEach']>): void;
}

class User implements Identifiable, Named {
    constructor(public name) {
        this.id = Utils.generateUUID();
    }
}
class Schedule implements Identifiable, Foreachable<string> {
    id!;
    private events = new Array<string>();

    forEach(...args) {
        return this.events.forEach(...args);
    }

}
```
## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
Duplicate #10570 / #6118 / #1373
